### PR TITLE
Try to handle expected error in record processor

### LINKED ISSUE #9420: Support logical transactions for command processing
We want to be able to reject the ProcessInstanceCreation command during certain circumstances, for example when the construction of the element scopes fails. When we reject the command, no changes should be made to the state. In addition, we may want to avoid writing events to the log stream that were created as part of the now rejected command.

To accommodate this, we should support logical transactions for our command processing.

- [x] We'll probably need to add an abstraction around transactions that includes rollback for a batch of records, state changes, and side-effects (or avoiding that those have happened until the transaction can be atomically committed).

  -> When an exception is thrown by a processor, the StreamProcessor rolls back the transaction. The `Engine` has a `onProcessingError` method which is then called with the failed command, the exception and a new ResultBuilder.

  - [ ] We need to pass the error to the corresponding processor so it handle it accordingly.

Please make sure to document how logical transactions work as part of this issue.

### LINKED ISSUE #9644: Reject create process instance command using logical transaction
**Description**

#9637 used a workaround to reject the Create Process Instance command when the engine fails to subscribe to the relevant catch events. The workaround is to throw an exception and let the ProcessingStateMachine handle the exception as an error. The workaround was necessary because of a lack of support for:

- #9420

As soon as logical transactions are supported, we should use the logical transaction to rollback the changes and reject the command.

### PR DESCRIPTION
**Pull request by korthout**
## Description

This PR offers a solution for
- #9420 

The idea is to pass the error to the specific `TypedRecordProcessor`, which can decide whether the error is `EXPECTED` or `UNEXPECTED`. 
- If the error is `UNEXPECTED`, we handle it as before by writing an Error event, etc.
- If the error is `EXPECTED`, we let the processor control how we reject the command and what else has to happen.

This PR shows how this could work by also providing a solution for
- #9644 

A test is added to verify the correct behavior: custom rejection + transaction rolled back. The test showed that a hidden bug was introduced (I managed to bisect this to 463366f4e9901aa6c001d37c022b75b774f4a009): the output was reset on error, but no longer when encapsulated by the ProcessingResultBuilder. A fix for this bug is also part of this PR.

## Related issues

closes #9420
closes #9644

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by korthout** [comment]
🤔 It seems that the changed test case `CreateProcessInstanceRejectionTest.shouldRejectCommandIfUnableToSubscribeToEvents` still fails after the main change because somehow the follow-up events are still written to the log. It's unclear to me how, but I guess we have a bug in resetting the result builder.

**Comment by korthout** [comment]
@Zelldon, @pihme, @saig0, and I discussed this concept. We've accepted the main concept, so I'll now continue to make this PR ready for review. For details please see this [slack comment](https://camunda.slack.com/archives/C03HW6MG3Q9/p1660037128878929?thread_ts=1659710745.506629&cid=C03HW6MG3Q9).

To recap: the current approach solves the problem, and is more future proof, than the alternative. Also, we can always iterate on the approach if needed.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java [hunk 1]
```diff
@@ -118,6 +121,17 @@ public boolean onCommand(
     return true;
   }
 
+  @Override
+  public ProcessingError tryHandleError(
+      final TypedRecord<ProcessInstanceCreationRecord> typedCommand, final Throwable error) {
+    if (error instanceof EventSubscriptionException exception) {
```
Comment (pihme) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:127]: :question: Is this always true? Or is it only true when we are in a "token move" context?

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingStateMachine.java [hunk 2]
```diff
@@ -326,6 +326,7 @@ private void errorHandlingInTransaction(final Throwable processingException) thr
           final long position = typedCommand.getPosition();
           final ProcessingResultBuilder processingResultBuilder =
               new DirectProcessingResultBuilder(context, position);
+          processingResultBuilder.reset();
```
Comment (pihme) [inline engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingStateMachine.java:328]: :wrench: Please add a TODO to revisit this line. I think Chris wants to remove the reset method altogether. And once we have the new implementation, which is based on buffered writers, the reset will no longer be necessary, as creating a new `ProcessingResultBuilder` will also create a new, empty buffer

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingStateMachine.java [hunk 3]
```diff
@@ -285,7 +285,6 @@ private void processCommand(final LoggedEvent command) {
     } catch (final UnrecoverableException unrecoverableException) {
       throw unrecoverableException;
     } catch (final Exception e) {
-      LOG.error(ERROR_MESSAGE_PROCESSING_FAILED_SKIP_EVENT, command, metadata, e);
```
Comment (pihme) [inline engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingStateMachine.java:288]: :x: please add to JavaDoc contract of `RecordProcessor` that implementing classes are responsible for logging errors. This is something that @deepthidevaki  will also have to consider for her implementation

**Review by pihme — APPROVED** [review]
Looks good. 

One mandatory question and one mandatory change. Please take a look

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java (hunk 1 above)
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:127]: I've clarified it in a comment, and git commit message

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingStateMachine.java (hunk 2 above)
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingStateMachine.java:329]: I've opened #10047 and referred to it in a todo

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingStateMachine.java (hunk 3 above)
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/streamprocessor/ProcessingStateMachine.java:288]: I've adjusted the Javadoc and added a contract

**Comment by korthout** [comment]
@pihme Please have another look. Especially the contract could use an additional round of review.

File: engine/src/main/java/io/camunda/zeebe/engine/api/RecordProcessor.java [hunk 4]
```diff
@@ -63,7 +63,22 @@
...
 
   /**
-   * Called by platform when a processing error occurred
+   * Called by platform when a processing error occurred.
+   *
+   * <p><em>Contract</em>
+   *
+   * <ul>
+   *   <li>Record will be a command
+   *   <li>Will be called if an uncaught exception is thrown in {@link #process(TypedRecord,
```
Comment (pihme) [inline engine/src/main/java/io/camunda/zeebe/engine/api/RecordProcessor.java:71]: :wrench: : Will also be called when an uncaught exception is thrown in ProcessingStateMachine during commit of the transaction

**Review by pihme — APPROVED** [review]
One minor correction

**Comment by korthout** [comment]
bors merge

**Pull request merged**